upgrade/upgrades: TestBackfillJobsInfoTable failed

upgrade/upgrades.TestBackfillJobsInfoTable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10014983?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10014983?buildTab=artifacts#/) on release-23.1 @ [a0a3b078882aa1f1a73f6683fd5fef0137626fed](https://github.com/cockroachdb/cockroach/commits/a0a3b078882aa1f1a73f6683fd5fef0137626fed):


```
=== RUN   TestBackfillJobsInfoTable
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable3025739298
    test_log_scope.go:79: use -show-logs to present logs inline
    backfill_jobs_info_table_migration_test.go:107: query '
        SELECT count(*) FROM system.jobs AS j, system.job_info AS i
        WHERE j.id = i.job_id AND (j.payload = i.value OR j.progress = i.value) AND (j.id >= 1 AND j.id <= 5)
        ': expected:
        14
        
        got:
        12
        
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable3025739298
--- FAIL: TestBackfillJobsInfoTable (32.91s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101374 upgrade/upgrades: TestBackfillJobsInfoTable failed [A-jobs C-test-failure O-robot T-jobs branch-release-23.1.0]
- #100995 upgrade/upgrades: TestBackfillJobsInfoTable failed [A-jobs C-test-failure O-robot T-jobs branch-master]
</p>
</details>
/cc @cockroachdb/sql-schema
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackfillJobsInfoTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27810

upgrade/upgrades.TestBackfillJobsInfoTable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10033878?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10033878?buildTab=artifacts#/) on release-23.1 @ [43347361e631585acba789669c886d5343276ac0](https://github.com/cockroachdb/cockroach/commits/43347361e631585acba789669c886d5343276ac0):


```
=== RUN   TestBackfillJobsInfoTable
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable309814320
    test_log_scope.go:79: use -show-logs to present logs inline
    backfill_jobs_info_table_migration_test.go:107: query '
        SELECT count(*) FROM system.jobs AS j, system.job_info AS i
        WHERE j.id = i.job_id AND (j.payload = i.value OR j.progress = i.value) AND (j.id >= 1 AND j.id <= 5)
        ': expected:
        14
        
        got:
        10
        
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable309814320
--- FAIL: TestBackfillJobsInfoTable (35.15s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101374 upgrade/upgrades: TestBackfillJobsInfoTable failed [A-jobs C-test-failure O-robot T-jobs branch-release-23.1.0]
- #100995 upgrade/upgrades: TestBackfillJobsInfoTable failed [A-jobs C-test-failure O-robot T-jobs branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackfillJobsInfoTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

upgrade/upgrades.TestBackfillJobsInfoTable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10243183?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10243183?buildTab=artifacts#/) on release-23.1 @ [6cf937278a19b59b50abdf78f0e257807367600d](https://github.com/cockroachdb/cockroach/commits/6cf937278a19b59b50abdf78f0e257807367600d):


```
=== RUN   TestBackfillJobsInfoTable
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable2878705094
    test_log_scope.go:79: use -show-logs to present logs inline
    backfill_jobs_info_table_migration_test.go:107: query '
        SELECT count(*) FROM system.jobs AS j, system.job_info AS i
        WHERE j.id = i.job_id AND (j.payload = i.value OR j.progress = i.value) AND (j.id >= 1 AND j.id <= 5)
        ': expected:
        14
        
        got:
        10
        
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable2878705094
--- FAIL: TestBackfillJobsInfoTable (32.02s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101374 upgrade/upgrades: TestBackfillJobsInfoTable failed [A-jobs C-test-failure O-robot T-jobs todo-deprecate.branch-release-23.1.0]
- #100995 upgrade/upgrades: TestBackfillJobsInfoTable failed [A-jobs C-test-failure O-robot T-jobs branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackfillJobsInfoTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

upgrade/upgrades.TestBackfillJobsInfoTable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10307865?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10307865?buildTab=artifacts#/) on release-23.1 @ [029955855b3243eecfbb7fcc0507ec663f7ddf6f](https://github.com/cockroachdb/cockroach/commits/029955855b3243eecfbb7fcc0507ec663f7ddf6f):


```
=== RUN   TestBackfillJobsInfoTable
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable3254874492
    test_log_scope.go:79: use -show-logs to present logs inline
    backfill_jobs_info_table_migration_test.go:107: query '
        SELECT count(*) FROM system.jobs AS j, system.job_info AS i
        WHERE j.id = i.job_id AND (j.payload = i.value OR j.progress = i.value) AND (j.id >= 1 AND j.id <= 5)
        ': expected:
        14
        
        got:
        12
        
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable3254874492
--- FAIL: TestBackfillJobsInfoTable (31.02s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100995 upgrade/upgrades: TestBackfillJobsInfoTable failed [A-jobs C-test-failure O-robot T-jobs branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackfillJobsInfoTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

upgrade/upgrades.TestBackfillJobsInfoTable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10483206?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10483206?buildTab=artifacts#/) on release-23.1 @ [8cfaf3b4731273cedaa9e81600a7b1778700b702](https://github.com/cockroachdb/cockroach/commits/8cfaf3b4731273cedaa9e81600a7b1778700b702):


```
=== RUN   TestBackfillJobsInfoTable
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable3319567719
    test_log_scope.go:79: use -show-logs to present logs inline
    backfill_jobs_info_table_migration_test.go:107: query '
        SELECT count(*) FROM system.jobs AS j, system.job_info AS i
        WHERE j.id = i.job_id AND (j.payload = i.value OR j.progress = i.value) AND (j.id >= 1 AND j.id <= 5)
        ': expected:
        14
        
        got:
        9
        
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable3319567719
--- FAIL: TestBackfillJobsInfoTable (32.52s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100995 upgrade/upgrades: TestBackfillJobsInfoTable failed [A-jobs C-test-failure O-robot T-jobs branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackfillJobsInfoTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

upgrade/upgrades.TestBackfillJobsInfoTable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10509150?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10509150?buildTab=artifacts#/) on release-23.1 @ [7cf4908e4d88fa1cc4127bbc4c4e327e2aed8b6d](https://github.com/cockroachdb/cockroach/commits/7cf4908e4d88fa1cc4127bbc4c4e327e2aed8b6d):


```
=== RUN   TestBackfillJobsInfoTable
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable2558230056
    test_log_scope.go:79: use -show-logs to present logs inline
    backfill_jobs_info_table_migration_test.go:107: query '
        SELECT count(*) FROM system.jobs AS j, system.job_info AS i
        WHERE j.id = i.job_id AND (j.payload = i.value OR j.progress = i.value) AND (j.id >= 1 AND j.id <= 5)
        ': expected:
        14
        
        got:
        10
        
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/194b1b11adb2bb86a3c51c4924595952/logTestBackfillJobsInfoTable2558230056
--- FAIL: TestBackfillJobsInfoTable (27.57s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100995 upgrade/upgrades: TestBackfillJobsInfoTable failed [A-jobs C-test-failure O-robot T-jobs branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackfillJobsInfoTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/upgrade/upgrades/BUILD.bazel
+++ b/pkg/upgrade/upgrades/BUILD.bazel
@@ -135,6 +135,8 @@ go_test(
     shard_count = 16,
     deps = [
         "//pkg/base",
+        "//pkg/ccl/backupccl",
+        "//pkg/ccl/changefeedccl",
         "//pkg/cloud/userfile",
         "//pkg/clusterversion",
         "//pkg/jobs",

--- a/pkg/upgrade/upgrades/backfill_job_info_table_migration_test.go
+++ b/pkg/upgrade/upgrades/backfill_job_info_table_migration_test.go
@@ -15,11 +15,15 @@ import (
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
+	_ "github.com/cockroachdb/cockroach/pkg/ccl/backupccl"
+	_ "github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl"
 	"github.com/cockroachdb/cockroach/pkg/clusterversion"
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/security/username"
 	"github.com/cockroachdb/cockroach/pkg/server"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/upgrade/upgrades"
@@ -53,25 +57,20 @@ func TestBackfillJobsInfoTable(t *testing.T) {
 	r := tc.Server(0).JobRegistry().(*jobs.Registry)
 	sqlDB := sqlutils.MakeSQLRunner(tc.ServerConn(0))
 	defer tc.Stopper().Stop(ctx)
-	jobspb.ForEachType(func(typ jobspb.Type) {
-		// The upgrade creates migration and schemachange jobs, so we do not
-		// need to create more. We should not override resumers for these job types,
-		// otherwise the upgrade will hang.
-		if typ != jobspb.TypeMigration && typ != jobspb.TypeSchemaChange {
-			r.TestingWrapResumerConstructor(typ, func(r jobs.Resumer) jobs.Resumer { return &fakeResumer{} })
-		}
-	}, false)
 
-	createJob := func(id jobspb.JobID, details jobspb.Details, progress jobspb.ProgressDetails) *jobs.Job {
+	createJob := func(id jobspb.JobID, details jobspb.Details, progress jobspb.ProgressDetails) {
 		defaultRecord := jobs.Record{
 			Details:  details,
 			Progress: progress,
 			Username: username.TestUserName(),
 		}
 
-		job, err := r.CreateJobWithTxn(ctx, defaultRecord, id, nil /* txn */)
+		var job *jobs.StartableJob
+		db := tc.Server(0).InternalDB().(*sql.InternalDB)
+		err := db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			return r.CreateStartableJobWithTxn(ctx, &job, id, txn, defaultRecord)
+		})
 		require.NoError(t, err)
-		return job
 	}
 
 	// Create a few different types of jobs.

